sql: TestSchemaChangeReverseMutations failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/a0b7cd4ebddf5ebc8f8c2119b119e57688f072f9

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestSchemaChangeReverseMutations PKG=github.com/cockroachdb/cockroach/pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=969388&tab=buildLog

```
I181016 12:36:53.454181 108165 sql/lease.go:315  publish (1 leases): desc=[{test 53 16}]
I181016 12:36:53.865056 108165 sql/lease.go:345  [n1,scExec] publish: descID=53 (test) version=18 mtime=2018-10-16 12:36:53.863309659 +0000 UTC
I181016 12:36:53.955236 108165 sql/lease.go:315  publish (1 leases): desc=[{test 53 17}]
I181016 12:36:53.976904 108750 sql/lease.go:877  new lease: 53("test") ver=18:1539693670.672605501,0, refcount=0
I181016 12:36:54.019884 108165 sql/lease.go:315  publish (1 leases): desc=[{test 53 17}]
I181016 12:36:54.098961 108165 sql/backfill.go:132  [n1,scExec] Running backfill for "test", v=18, m=5
I181016 12:36:54.303246 108133 storage/replica_proposal.go:212  [n1,s1,r5/1:/System/ts{d-e}] new range lease repl=(n1,s1):1 seq=3 start=1539693403.397612557,0 epo=1 pro=1539693414.297802820,0 following repl=(n1,s1):1 seq=2 start=1539693403.397612557,0 exp=1539693412.447715340,0 pro=1539693403.447847978,0
I181016 12:36:54.348601 108165 sql/lease.go:345  [n1,scExec] publish: descID=53 (test) version=19 mtime=2018-10-16 12:36:54.346664967 +0000 UTC
I181016 12:36:54.478938 108165 sql/event_log.go:126  [n1,scExec] Event: "finish_schema_change", target: 53, info: {MutationID:5}
I181016 12:36:54.537191 107868 sql/schema_changer_test.go:1767  SucceedsSoon: 1 mutations remaining
I181016 12:36:54.576519 108165 sql/lease.go:315  publish (1 leases): desc=[{test 53 18}]
I181016 12:36:54.583863 108793 sql/lease.go:877  new lease: 53("test") ver=19:1539693663.173692229,0, refcount=0
I181016 12:36:54.629890 108165 sql/lease.go:315  publish (1 leases): desc=[{test 53 18}]
I181016 12:36:54.879956 108165 sql/lease.go:345  [n1,scExec] publish: descID=53 (test) version=20 mtime=2018-10-16 12:36:54.87841889 +0000 UTC
I181016 12:36:54.923463 108165 sql/lease.go:315  publish (1 leases): desc=[{test 53 19}]
I181016 12:36:54.924935 108797 sql/lease.go:877  new lease: 53("test") ver=20:1539693711.125095242,0, refcount=0
I181016 12:36:54.968340 108165 sql/backfill.go:132  [n1,scExec] Running backfill for "test", v=20, m=6
I181016 12:36:55.192760 108165 sql/lease.go:345  [n1,scExec] publish: descID=53 (test) version=21 mtime=2018-10-16 12:36:55.190844125 +0000 UTC
I181016 12:36:55.311468 108165 sql/event_log.go:126  [n1,scExec] Event: "finish_schema_change", target: 53, info: {MutationID:6}
I181016 12:36:55.368868 108948 sql/lease.go:877  new lease: 53("test") ver=21:1539693665.963444366,0, refcount=0
I181016 12:36:55.387733 108165 sql/lease.go:315  publish (1 leases): desc=[{test 53 20}]
I181016 12:36:55.544648 107868 sql/schema_changer_test.go:1767  SucceedsSoon: 1 gc mutations remaining
I181016 12:36:55.609463 108165 sql/lease.go:345  [n1,scExec] publish: descID=53 (test) version=22 mtime=2018-10-16 12:36:55.607695096 +0000 UTC
I181016 12:36:55.744490 108998 sql/lease.go:877  new lease: 53("test") ver=22:1539693732.945594691,0, refcount=0
I181016 12:36:55.753812 108165 sql/lease.go:315  publish (1 leases): desc=[{test 53 21}]
I181016 12:36:55.803934 108165 sql/lease.go:315  publish (1 leases): desc=[{test 53 21}]
I181016 12:36:56.612276 107868 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] transport racer
1      [async] storage.consistencyChecker: processing replica
1      [async] storage.Replica: checking consistency
1      [async] closedts-rangefeed-subscriber
I181016 12:36:56.614533 107868 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] transport racer
1      [async] storage.consistencyChecker: processing replica
1      [async] closedts-rangefeed-subscriber
I181016 12:36:56.615382 107868 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] transport racer
1      [async] closedts-rangefeed-subscriber
I181016 12:36:56.617555 108110 kv/transport_race.go:113  transport race promotion: ran 62 iterations on up to 4084 requests
I181016 12:36:56.620412 107868 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-rangefeed-subscriber
--- FAIL: TestSchemaChangeReverseMutations (13.48s)
	schema_changer_test.go:1782: pq: batch timestamp 1539693416.550241344,0 must be after GC threshold 1539693416.551361159,0

```

--- a/pkg/sql/schema_changer_test.go
+++ b/pkg/sql/schema_changer_test.go
@@ -1701,10 +1701,6 @@ CREATE TABLE t.test (k INT PRIMARY KEY, v INT);
 	}
 
 	tableDesc := sqlbase.GetTableDescriptor(kvDB, "t", "test")
-	// Add a zone config for the table.
-	if _, err := addImmediateGCZoneConfig(sqlDB, tableDesc.ID); err != nil {
-		t.Fatal(err)
-	}
 
 	testCases := []struct {
 		sql    string
@@ -1758,46 +1754,39 @@ CREATE TABLE t.test (k INT PRIMARY KEY, v INT);
 		t.Fatalf("e = %d, v = %d", e, len(tableDesc.Mutations))
 	}
 
-	// Enable async schema change processing.
+	// Enable async schema change processing for purged schema changes.
 	atomic.StoreUint32(&enableAsyncSchemaChanges, 1)
 
-	// Wait until all the mutations have been processed.
-	var rows *gosql.Rows
 	expectedCols := []string{"k", "b", "d"}
+	// Wait until all the mutations have been processed.
 	testutils.SucceedsSoon(t, func() error {
-		// Read table descriptor.
 		tableDesc = sqlbase.GetTableDescriptor(kvDB, "t", "test")
 		if len(tableDesc.Mutations) > 0 {
 			return errors.Errorf("%d mutations remaining", len(tableDesc.Mutations))
 		}
-		if len(tableDesc.GCMutations) > 0 {
-			return errors.Errorf("%d gc mutations remaining", len(tableDesc.GCMutations))
-		}
-
-		// Verify that t.public.test has the expected data. Read the table data while
-		// ensuring that the correct table lease is in use.
-		var err error
-		rows, err = sqlDB.Query(`SELECT * from t.test`)
-		if err != nil {
-			t.Fatal(err)
-		}
-		cols, err := rows.Columns()
-		if err != nil {
-			t.Fatal(err)
-		}
-
-		// Ensure that sql is using the correct table lease.
-		if len(cols) != len(expectedCols) {
-			defer rows.Close()
-			return errors.Errorf("incorrect columns: %v, expected: %v", cols, expectedCols)
-		}
-		if cols[0] != expectedCols[0] || cols[1] != expectedCols[1] {
-			t.Fatalf("incorrect columns: %v", cols)
-		}
 		return nil
 	})
 
+	// Verify that t.public.test has the expected data. Read the table data while
+	// ensuring that the correct table lease is in use.
+	rows, err := sqlDB.Query(`SELECT * from t.test`)
+	if err != nil {
+		t.Fatal(err)
+	}
 	defer rows.Close()
+	cols, err := rows.Columns()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Ensure that sql is using the correct table lease.
+	if len(cols) != len(expectedCols) {
+		t.Fatalf("incorrect columns: %v, expected: %v", cols, expectedCols)
+	}
+	if cols[0] != expectedCols[0] || cols[1] != expectedCols[1] {
+		t.Fatalf("incorrect columns: %v", cols)
+	}
+
 	// rows contains the data; verify that it's the right data.
 	vals := make([]interface{}, len(expectedCols))
 	for i := range vals {
@@ -1847,7 +1836,7 @@ CREATE TABLE t.test (k INT PRIMARY KEY, v INT);
 
 	// Check that the index on b eventually goes live even though a schema
 	// change in front of it in the queue got purged.
-	rows, err := sqlDB.Query(`SELECT * from t.test@test_b_key`)
+	rows, err = sqlDB.Query(`SELECT * from t.test@test_b_key`)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -1867,14 +1856,27 @@ CREATE TABLE t.test (k INT PRIMARY KEY, v INT);
 		t.Fatal("SELECT over index 'foo' works")
 	}
 
+	if err := sqlutils.RunScrub(sqlDB, "t", "test"); err != nil {
+		t.Fatal(err)
+	}
+
+	// Add immediate GC TTL to allow index creation purge to complete.
+	if _, err := addImmediateGCZoneConfig(sqlDB, tableDesc.ID); err != nil {
+		t.Fatal(err)
+	}
+
+	testutils.SucceedsSoon(t, func() error {
+		tableDesc = sqlbase.GetTableDescriptor(kvDB, "t", "test")
+		if len(tableDesc.GCMutations) > 0 {
+			return errors.Errorf("%d gc mutations remaining", len(tableDesc.GCMutations))
+		}
+		return nil
+	})
+
 	ctx := context.TODO()
 
 	// Check that the number of k-v pairs is accurate.
 	if err := checkTableKeyCount(ctx, kvDB, 3, maxValue); err != nil {
-		t.Fatal(err)
-	}
-
-	if err := sqlutils.RunScrub(sqlDB, "t", "test"); err != nil {
 		t.Fatal(err)
 	}
 

